### PR TITLE
Additional refactoring of hash functions

### PR DESCRIPTION
**Pull request by bdice**
Additional work related to #10081.

This is breaking because it reorganizes several public names/namespaces.

Summary of changes in this PR:
- The `cudf` namespace now wraps the contents of `hash_functions.cuh`, and some public names are now classified as `detail` APIs.
- `SparkMurmurHash3_32` has been updated to align with the design and naming conventions of `MurmurHash3_32`

### CODE REVIEW

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (bdice) [inline cpp/include/cudf/detail/utilities/hash_functions.cuh:597]: Should the `default_hash` be in a `detail` namespace or not? I put the hash function implementations like `MurmurHash3_32` in a `detail` namespace.

File: python/cudf/cudf/core/resample.py
Comment (bdice) [inline python/cudf/cudf/core/resample.py:1]: Is this copyright correct? Should we rewrite this file's copyright header? @rapidsai/cudf-python-codeowners

Comment (bdice) [reply python/cudf/cudf/core/resample.py:1]: Looks like this is #9701 (@shwina). Unfortunately that PR has no description to indicate why the copyright was updated.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (jrhemstad) [inline cpp/include/cudf/detail/utilities/hash_functions.cuh:244]: Use `[[fallthrough]]` here

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (jrhemstad) [inline cpp/include/cudf/detail/utilities/hash_functions.cuh:374]: Future PR, but this would be more efficient to round up to the nearest 4B aligned address and do a 4B load and then shift the bits as appropriate to get them back in the expected order.

Comment (bdice) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:374]: You'd need to do _two_ 4B loads and shift both results together if you were reading 4B aligned addresses, right? This is meant to handle the case where the 4 bytes are not aligned (could cross 4B alignment boundaries).

**Review by jrhemstad — APPROVED** [review]

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (jrhemstad) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:374]: Yep, you'd have to do two when it's not already aligned.

Comment (jrhemstad) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:374]: We have code for this sitting around in various places.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (bdice) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh]: Nice. Fixed in 0a170183e3.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (bdice) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:374]: Ok great. I'll figure that out in a future PR -- there are still other changes I have in mind, but this PR is needed to unblock other work so I'm trying to limit its scope.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (vyasr) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:597]: My recollection is that the `default_hash` is only used as the default hash function for various internal hash tables (join, groupby, etc), so it's only necessary for developers. As such, `detail` seems appropriate.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (bdice) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:394]: This is in the `SparkMurmurHash3_32` class, so it adheres to Spark's conventions. The `MurmurHash3_32` class has a different approach. At some later point I may reorganize these classes to use inheritance or CRTP or something to reduce the duplication of logic.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (bdice) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:44]: I'll leave this unchanged.

File: cpp/src/text/subword/load_merges_file.cu
Comment (bdice) [reply cpp/src/text/subword/load_merges_file.cu:121]: I agree this is awkward / undesirable. This may be resolved or improved by #10401.

Comment (bdice) [reply cpp/src/text/subword/load_merges_file.cu:121]: Actually, no. The issue is that we need to provide a stream and that requires re-supplying the other defaulted arguments. https://github.com/NVIDIA/cuCollections/blob/fb58a38701f1c24ecfe07d8f1f208bbe80930da5/include/cuco/static_map.cuh#L224-L231

**Comment by bdice** [comment]
@gpucibot merge

File: cpp/src/text/subword/load_merges_file.cu
Comment (vyasr) [reply cpp/src/text/subword/load_merges_file.cu:121]: Sure, but IMO it's a design issue in cuco if supplying a stream while using the default hash requires pulling the default hash out of a detail namespace.

Comment (bdice) [reply cpp/src/text/subword/load_merges_file.cu:121]: I agree with that. (Sorry, the "Actually, no" was about whether #10401 would improve this situation. It would not.)

Comment (vyasr) [reply cpp/src/text/subword/load_merges_file.cu:121]: Got it. I think in the long run we probably don't want to be able to provide the hash function and equality comparator as template parameters of these methods, but rather as parameters of the constructor. It doesn't really make sense to be able to insert and query with different ones. Unfortunately we currently abuse this ability in libcudf, so I don't think removing it is feasible in the short term, but in the longer term getting rid of this would make it easier to provide streams without having this problem since the hash/equality operators would be defined on construction and the user wouldn't need to provide those templates unless they wanted to override the default.

**Pull request merged**
